### PR TITLE
[Provider] Openrouter, Lingyi, Moonshot and Zhipu

### LINKED ISSUE #324: [Provider]: Openrouter
https://openrouter.ai/

### LINKED ISSUE #325: [Provider]: Zhipu
https://www.zhipuai.cn/en/

### LINKED ISSUE #326: [Provider]: Lingyi
https://platform.lingyiwanwu.com/apikeys

### LINKED ISSUE #327: [Provider]: Moonshot
https://platform.moonshot.cn/console/api-keys

### PR DESCRIPTION
**Pull request by sternelee**
Support [openrouter](https://openrouter.ai/).ai, [lingyiwanwu.com](https://platform.lingyiwanwu.com/) and [moonshot.cn](https://www.moonshot.cn/), [bigmodel.cn](https://open.bigmodel.cn/) AI provider

### CODE REVIEW

File: pnpm-lock.yaml
Comment (VisargD) [inline pnpm-lock.yaml:1]: Can you please remove this file as its not required for the current project setup and is out of scope for the PR.

File: wrangler.toml
Comment (VisargD) [inline wrangler.toml:1]: Can you please revert the changes of this file. These are not required for the PR and we have kept toml file open ended so user can change it or their deployment.

**Comment by VisargD** [comment]
Hey @sternelee - Can you please also run `npm run format` to execute prettier formatting of the project. The PR check is failing due to it.

File: src/providers/lingyi/chatComplete.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      })),
+      usage: {
+        prompt_tokens: response.usage?.prompt_tokens || 0,
+        completion_tokens: response.usage?.completion_tokens || 0,
+        total_tokens: response.usage?.total_tokens || 0,
+      },
+    };
+  }
+
+  return {
```
Comment (VisargD) [inline src/providers/lingyi/chatComplete.ts]: You can use `generateInvalidProviderResponseError` to generate the invalid response error message for any provider. Applicable for all the provider integrated in this PR.
```
return generateInvalidProviderResponseError(response, LINGYI);
```

You can check other providers integrations for example.

Comment (VisargD) [inline src/providers/lingyi/chatComplete.ts]: You can use `generateInvalidProviderResponseError` to generate the invalid response error message for any provider. Applicable for all the provider integrated in this PR.
```
return generateInvalidProviderResponseError(response, LINGYI);
```

You can check other providers integrations for example.

File: src/providers/lingyi/chatComplete.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    finish_reason: string | null;
+  }[];
+}
+
+export const LingYiChatCompleteResponseTransform: (
+  response: LingYiChatCompleteResponse | LingYiErrorResponse,
+  responseStatus: number,
+) => ChatCompletionResponse | ErrorResponse = (response, responseStatus) => {
+  if ("error" in response && responseStatus !== 200) {
+    return {
```
Comment (VisargD) [inline src/providers/lingyi/chatComplete.ts]: You can use `generateErrorResponse` to generate the provider error response. Applicable for all the provider integrated in this PR.
```
return generateErrorResponse(
  {
    message: response.error.message,
    type: response.error.type,
    param: null,
    code: response.error.code?.toString() || null,
  },
  LINGYI
);
```

**Review by VisargD — CHANGES_REQUESTED** [review]
Hey @sternelee - Thanks for the PR! I have added first set of comments for the PR. Please check and let me know if you need any help with them

File: wrangler.toml
Comment (sternelee) [reply wrangler.toml:1]: got it

File: src/providers/lingyi/chatComplete.ts (hunk 1 above)
Comment (sternelee) [reply src/providers/lingyi/chatComplete.ts]: Okay, I just noticed that the logical structure has been updated

File: wrangler.toml
Comment (VisargD) [inline wrangler.toml:1]: Sorry for the nitpick: I think there was one new line removed. If its possible please completely remove this file from the diff.

File: src/providers/openrouter/chatComplete.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  return generateInvalidProviderResponseError(response, OPENROUTER);
+};
+
+export const OpenrouterChatCompleteStreamChunkTransform: (
+  response: string
+) => string = (responseChunk) => {
+  let chunk = responseChunk.trim();
+  chunk = chunk.replace(/^data: /, '');
+  chunk = chunk.trim();
+  if (chunk === '[DONE]') {
```
Comment (VisargD) [inline src/providers/openrouter/chatComplete.ts:128]: openrouter also sends a `:OPENROUTER PROCESSING` event as a part of stream. Can you please add a check just like `[DONE] ` but return `undefined` in that case. Example:
```
if (chunk === ': OPENROUTER PROCESSING') {
    return undefined;
  }
```

Comment (VisargD) [reply src/providers/openrouter/chatComplete.ts:128]: Screenshot 2024-04-10 at 18 14 56

File: wrangler.toml
Comment (sternelee) [reply wrangler.toml:1]: Oh, this is the code style of my editor, I don't think there is a need to leave a blank line at the end

File: src/providers/openrouter/chatComplete.ts (hunk 3 above)
Comment (sternelee) [reply src/providers/openrouter/chatComplete.ts:128]: you are right, I think we just return the same value. Indeed, it is feasible for me to use `eventsource-parser` in other projects. Thanks for your good work.

Comment (VisargD) [reply src/providers/openrouter/chatComplete.ts:128]: Returning the same value will break OpenAI/other compatible SDKs as they expect a different structure. We will have to return undefined in this case so that event is skipped from being sent to the clients.

**Comment by sternelee** [comment]
@VisargD How do I modify the body structure？I want to adapted for reka.ai and coze.com bot

File: src/providers/openrouter/chatComplete.ts [hunk 4]
```diff
@@ -81,6 +81,7 @@ export const OpenrouterChatCompleteResponseTransform: (
   response: OpenrouterChatCompleteResponse | OpenrouterErrorResponse,
   responseStatus: number
 ) => ChatCompletionResponse | ErrorResponse = (response, responseStatus) => {
+  console.log('response:', response);
```
Comment (VisargD) [inline src/providers/openrouter/chatComplete.ts]: Sorry for the nitpick: Can you please remove console logs from the PR?

Comment (sternelee) [reply src/providers/openrouter/chatComplete.ts]: sorry for my negligence

File: src/providers/openrouter/chatComplete.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+) => string = (responseChunk) => {
+  let chunk = responseChunk.trim();
+  chunk = chunk.replace(/^data: /, '');
+  chunk = chunk.trim();
+  if (chunk === '[DONE]') {
+    return `data: ${chunk}\n\n`;
+  }
+  if (chunk.includes('OPENROUTER PROCESSING')) {
+    chunk = JSON.stringify({
+      id: 0,
```
Comment (unsync) [inline src/providers/openrouter/chatComplete.ts]: @sternelee according to the OpenAI api signature, the id is a string : 

https://platform.openai.com/docs/api-reference/chat/object

Using an integer causes some sdk to throw an error for these chunks, could you set a string value like this ?

```
id: `${Date.now()}`,
```

cc @VisargD

File: src/providers/openrouter/chatComplete.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      model: '',
+      object: 'chat.completion.processing',
```
Comment (unsync) [inline src/providers/openrouter/chatComplete.ts]: @sternelee Similar issue here, the OpenAI signature ask for `chat.completion.chunk` in all chunks :

https://platform.openai.com/docs/api-reference/chat/streaming#chat/streaming-object

**Comment by VisargD** [comment]
Looks good to me. I will merge the PR in the next release. Thanks!

**Comment by VisargD** [comment]
Hey @sternelee - Looks like PR was updated and it shows 0 files changed in the diff now. Can you please check this once?

**Comment by sternelee** [comment]
> Hey @sternelee - Looks like PR was updated and it shows 0 files changed in the diff now. Can you please check this once?

oh, I don’t know why it automatically sync with the upstream. I have already pushed force it again.

**Review by VisargD — APPROVED** [review]

**Comment by VisargD** [comment]
Closes #327 
Closes #326 
Closes #325 
Closes #324
